### ISSUE TITLE
Exception thrown in request iterator causes response iterator to hang forever (Python)

### ISSUE THREAD

**Issue by arthur-tacca** [grpc/grpc#8454]
I have a server with a streaming request and streaming response RPC, and a client for it, both written in Python 3.5.2 on Ubuntu 16 using gRPC 1.0.0. In the client, if the request iterator I give to gRPC throws an exception (thrown by _my_ code), then the iterator over response objects hangs forever. I see a stack trace printed to the console when this happens:

```
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/arthur/myenv/lib/python3.5/site-packages/grpc/_channel.py", line 184, in consume_request_iterator
    for request in request_iterator:
  File "/home/arthur/src/my_project/my_module.py", line 159, in my_request_iterator
    json_format.Parse(json.dumps(foo), bar)
  ...
  < I have removed some extra lines of irrelevant trace here >
```

Ideally, I'd prefer the exception object to be caught in gRPC's thread and then rethrown in the response iterator (with the stack trace preserved) so that my code can handle it. Failing that, I'd like to see some other exception, or really anything other than hanging forever.

Edit: I just tested gRPC 1.0.1rc1 and it has the same issue.

Edit 2: This is quite easy to workaround in user code (depending on how it is structured): put a try/except block at the top level of your request generator, record the exception somewhere (and don't re-raise it!), then check for it after iterating over responses.

**Comment by nathanielmanistaatgoogle** [comment]
@thunderboltsid: this isn't quite a duplicate of, but it certainly goes hand-in-hand with, [issue 8231](https://github.com/grpc/grpc/issues/8231). I'm not yet convinced that re-raising the exception from the response iterator is the right remedy; I'm open to something as simple as logging the error and cancelling the RPC. Raising a _different_ exception from the response iterator, along with logging the original exception, strikes me as perhaps the best resolution.

@arthur-tacca: thank you for the report and for trying out 1.0.1rc1!

**Comment by nathanielmanistaatgoogle** [comment]
The fix for this landed in [pull request 8686](https://github.com/grpc/grpc/pull/8686) and will be included in whatever release comes after `1.0.4` (likely `1.0.5` or `1.1.0`).
